### PR TITLE
[TOPIC-GPIO] gpio: Update imx gpio driver to use new gpio api

### PR DESCRIPTION
**Pull request by stanislav-poboril**
Updates the NXP i.MX gpio driver and all associated boards to use new
device tree compatible gpio configuration flags. Implements new port
get/set/clear/toggle and pin_interrupt_configure functions recently
added to the gpio api.

Assumes the gpio api layer handles translating logical flags to physical
flags.

Tested with:
- samples/basic/blinky
- samples/basic/button
- tests/drivers/gpio/gpio_api_1pin
- tests/drivers/gpio/gpio_basic_api

On boards:
- udoo_neo_full_m4

The udoo_neo_full_m4 board does not have a user button, so I temporarily modified the samples/basic/button code and board dts file to use second pin to simulate the button.

The test tests/drivers/gpio/gpio_api_1pin skips most of the tests, since it relies on pin to be configured as both input/output at the same time. i.MX gpio can be either input or output. However it is possible to read back the data from the output pins, so I temporarily modified the code to be able to test some of the cases.

I don't have colibri_imx7d_m4 board, only verified that the samples/basic/blinky and samples/basic/button examples build. The test tests/drivers/gpio/gpio_api_1pin compiles but does not link since it does not fit into board's flash.

### CODE REVIEW

**Comment by galak** [comment]
Are you able to test with the `tests/drivers/gpio/gpio_basic_api`.  It will require a wire to loop back a pair of pins.  This test covers more so feel pretty confident if we're able to pass it.

**Review by galak — DISMISSED** [review]
Should convert boards/arm/96b_meerkat96/96b_meerkat96.dts & boards/arm/warp7_m4/warp7_m4.dts

**Comment by stanislav-poboril** [comment]
> Are you able to test with the `tests/drivers/gpio/gpio_basic_api`. It will require a wire to loop back a pair of pins. This test covers more so feel pretty confident if we're able to pass it.

Updated the code, added the udoo_neo_full_m4 board configuration and pin mux code into the `tests/drivers/gpio/gpio_basic_api` sample. It is passing now.

> Should convert boards/arm/96b_meerkat96/96b_meerkat96.dts & boards/arm/warp7_m4/warp7_m4.dts

Done.

File: drivers/gpio/gpio_imx.c [hunk 1]
```diff
@@ -106,46 +117,162 @@ static int imx_gpio_read(struct device *dev,
...
+	u32_t icr_val;
+	u8_t shift;
+
+	if (((base->GDIR & BIT(pin)) != 0U)
+	    && (mode != GPIO_INT_MODE_DISABLED)) {
+		/* Interrupt on output pin not supported */
+		return -ENOTSUP;
+	}
+
+	if ((mode == GPIO_INT_MODE_LEVEL) && (trig == GPIO_INT_TRIG_BOTH)) {
```
Comment (mnkp) [inline drivers/gpio/gpio_imx.c]: I'll add this check as an assert to GPIO API. Let's remove it from the driver.

**Review by mnkp — CHANGES_REQUESTED** [review]

**Review by mnkp — APPROVED** [review]

**Comment by carlescufi** [comment]
@stanislav-poboril 
The topic branch has been rebased. Could you please rebase this PR against it?

**Comment by stanislav-poboril** [comment]
> 
> 
> @stanislav-poboril
> The topic branch has been rebased. Could you please rebase this PR against it?

Rebased and retested.

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**